### ISSUE TITLE
Receiving ClosedChannelException instead of actual error

### ISSUE THREAD

**Issue by ejona86** [grpc/grpc-java#1330]
We aren't reporting a helpful error in very simple connectivity cases, like the server port is closed. Here is what it looks like when using the interop client against a non-existent server: 

```
$ ./build/install/grpc-interop-testing/bin/test-client
Running test empty_unary
...
Exception in thread "main" io.grpc.StatusRuntimeException: UNAVAILABLE
        at io.grpc.Status.asRuntimeException(Status.java:431)
        at io.grpc.stub.ClientCalls.getUnchecked(ClientCalls.java:157)
        at io.grpc.stub.ClientCalls.blockingUnaryCall(ClientCalls.java:106)
        at io.grpc.testing.integration.TestServiceGrpc$TestServiceBlockingStub.emptyCall(TestServiceGrpc.java:212)
        at io.grpc.testing.integration.AbstractTransportTest.emptyUnary(AbstractTransportTest.java:164)
        at io.grpc.testing.integration.TestServiceClient.runTest(TestServiceClient.java:220)
        at io.grpc.testing.integration.TestServiceClient.run(TestServiceClient.java:209)
        at io.grpc.testing.integration.TestServiceClient.main(TestServiceClient.java:80)
Caused by: java.nio.channels.ClosedChannelException
```

It seems the problem may also impact other, much harder to diagnose, issues like client certificate validation (like in #1327).

**Comment by ejona86** [comment]
#1513 makes it obvious I didn't get all the cases. We need to find the missing cases that cause ClosedChannelException and then revert #1516.

**Comment by ejona86** [comment]
I saw a flake of this with #1594
```
io.grpc.netty.NettyTransportTest > ping_afterTermination FAILED
    java.lang.AssertionError: expected:<Status{code=UNAVAILABLE, description=null, cause=null}> but was:<Status{code=UNKNOWN, description=null, cause=java.nio.channels.ClosedChannelException}>
```

**Comment by buchgr** [comment]
With current master I get

```
Jakobs-MBP-4:interop-testing buchgr$ ./build/install/grpc-interop-testing/bin/test-client --use_tls=false
Jun 12, 2016 5:03:32 PM io.grpc.internal.ManagedChannelImpl <init>
INFO: [ManagedChannelImpl@239963d8] Created with target directaddress:///localhost/127.0.0.1:8080
Running test empty_unary
Jun 12, 2016 5:03:32 PM io.grpc.internal.ManagedChannelImpl maybeTerminateChannel
INFO: [ManagedChannelImpl@239963d8] Terminated
Exception in thread "main" io.grpc.StatusRuntimeException: UNAVAILABLE: Terminated fail fast stream.
	at io.grpc.Status.asRuntimeException(Status.java:534)
	at io.grpc.stub.ClientCalls.getUnchecked(ClientCalls.java:208)
	at io.grpc.stub.ClientCalls.blockingUnaryCall(ClientCalls.java:141)
	at io.grpc.testing.integration.TestServiceGrpc$TestServiceBlockingStub.emptyCall(TestServiceGrpc.java:356)
	at io.grpc.testing.integration.AbstractInteropTest.emptyUnary(AbstractInteropTest.java:174)
	at io.grpc.testing.integration.TestServiceClient.runTest(TestServiceClient.java:210)
	at io.grpc.testing.integration.TestServiceClient.run(TestServiceClient.java:198)
	at io.grpc.testing.integration.TestServiceClient.main(TestServiceClient.java:83)
Caused by: io.grpc.StatusException: INTERNAL: Connection closed with unknown cause
	at io.grpc.Status.asException(Status.java:551)
	at io.grpc.internal.DelayedClientTransport$3.run(DelayedClientTransport.java:327)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Shutting down
```

```
Jakobs-MBP-4:interop-testing buchgr$ ./build/install/grpc-interop-testing/bin/test-client
Jun 12, 2016 5:04:55 PM io.grpc.internal.ManagedChannelImpl <init>
INFO: [ManagedChannelImpl@239963d8] Created with target directaddress:///localhost/127.0.0.1:8080
Running test empty_unary
Jun 12, 2016 5:04:55 PM io.grpc.internal.ManagedChannelImpl maybeTerminateChannel
INFO: [ManagedChannelImpl@239963d8] Terminated
Exception in thread "main" io.grpc.StatusRuntimeException: UNAVAILABLE: Terminated fail fast stream.
	at io.grpc.Status.asRuntimeException(Status.java:534)
	at io.grpc.stub.ClientCalls.getUnchecked(ClientCalls.java:208)
	at io.grpc.stub.ClientCalls.blockingUnaryCall(ClientCalls.java:141)
	at io.grpc.testing.integration.TestServiceGrpc$TestServiceBlockingStub.emptyCall(TestServiceGrpc.java:356)
	at io.grpc.testing.integration.AbstractInteropTest.emptyUnary(AbstractInteropTest.java:174)
	at io.grpc.testing.integration.TestServiceClient.runTest(TestServiceClient.java:210)
	at io.grpc.testing.integration.TestServiceClient.run(TestServiceClient.java:198)
	at io.grpc.testing.integration.TestServiceClient.main(TestServiceClient.java:83)
Caused by: io.grpc.StatusException: UNAVAILABLE
	at io.grpc.Status.asException(Status.java:551)
	at io.grpc.internal.DelayedClientTransport$3.run(DelayedClientTransport.java:327)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
       ...
Caused by: javax.net.ssl.SSLHandshakeException: General OpenSslEngine problem
	at io.netty.handler.ssl.OpenSslContext$AbstractCertificateVerifier.verify(OpenSslContext.java:466)
	at org.apache.tomcat.jni.SSL.readFromSSL(Native Method)
	at io.netty.handler.ssl.OpenSslEngine.readPlaintextData(OpenSslEngine.java:365)
        ...
Caused by: java.security.cert.CertificateException: No subject alternative DNS name matching localhost found.
	at sun.security.util.HostnameChecker.matchDNS(HostnameChecker.java:204)
	at sun.security.util.HostnameChecker.match(HostnameChecker.java:95)
	at sun.security.ssl.X509TrustManagerImpl.checkIdentity(X509TrustManagerImpl.java:455)
        ...
Shutting down
```

**Comment by ejona86** [comment]
@buchgr, the first error still has "INTERNAL: Connection closed with unknown cause". We'd like to add better info. The second error looks like the server was running but TLS misconfigured somewhere (probably the client needs `--server_host_override=foo.test.google.fr`)

**Comment by buchgr** [comment]
@ejona86 I see. Will check. Yes, I misconfigured TLS on purpose cause of

> It seems the problem may also impact other, much harder to diagnose, issues like client certificate validation (like in #1327).

**Comment by ejona86** [comment]
@buchgr, ah, gotcha. https://github.com/grpc/grpc-java/issues/1330#issuecomment-191988166 is the interesting comment in that case. Note that I mention a commit to revert which is working around the issue for our tests.

**Comment by buchgr** [comment]
@ejona86 

As mentioned yesterday in the standup meeting, I investigated the `INTERNAL: Connection closed with unknown cause` error message. What's happening in Netty is that [SocketChannel.read(ByteBuffer)](https://github.com/netty/netty/blob/4.1/buffer/src/main/java/io/netty/buffer/PooledUnsafeDirectByteBuf.java#L288) returns `-1`, which according to the [docs ](https://docs.oracle.com/javase/7/docs/api/java/nio/channels/SocketChannel.html#read(java.nio.ByteBuffer)) means EOF. This then invokes [closeOnRead](https://github.com/netty/netty/blob/4.1/transport/src/main/java/io/netty/channel/nio/AbstractNioByteChannel.java#L131) which then just calls [Channel.close(voidPromise())](https://github.com/netty/netty/blob/4.1/transport/src/main/java/io/netty/channel/nio/AbstractNioByteChannel.java#L76). This then completes the `Channels` `ChannelFuture`, where gRPC is setting the Status to [Connection closed with unknown cause](https://github.com/grpc/grpc-java/blob/master/netty/src/main/java/io/grpc/netty/NettyClientTransport.java#L190).

I don't know whether we can get a better error message from Netty in that case. Also note, that in that case Netty would swallow a [ClosedChannelException](https://github.com/netty/netty/blob/4.1/buffer/src/main/java/io/netty/buffer/PooledUnsafeDirectByteBuf.java#L289).

There are a bunch of other ways we might notice that a Channel is no longer open, including on write, read with Exception, and the channel being closed after `Selector.select()` marking the `SelectionKey` invalid.

**Comment by buchgr** [comment]
I would therefore suggest to revert #1516 and see if that will make tests fail due to `ClosedChannelException` again?

**Comment by ejona86** [comment]
If it is actually EOF, that should be fine, because the server will have generally sent a GOAWAY before the EOF. In the case I was seeing earlier though, no connection should have been made to begin with. I'm pretty reliably getting "Connection refused" when trying now, so maybe it is fixed now.

I'm fine with reverting #1516 and giving it another go. But let's wait until after we release 0.15 so that we have some time to see if it reappears for us.

**Comment by buchgr** [comment]
@ejona86 
The scenario I am describing happens on the qps client when killing the server with `kill -9`. So no `GOAWAY` has been sent in that case. 

On the interop, hello world and qps client I am also getting reliably "Connection refused" if trying to connect to a closed port.

I ll open a PR to revert once 0.15 is released.

**Comment by ejona86** [comment]
I just saw a flake at https://travis-ci.org/grpc/grpc-java/jobs/140586965 (since rebuilt):
```
io.grpc.netty.NettyTransportTest > ping_afterTermination FAILED
    java.lang.AssertionError: expected:<Status{code=UNAVAILABLE, description=null, cause=null}> but was:<Status{code=UNKNOWN, description=null, cause=java.nio.channels.ClosedChannelException}>
```

**Comment by ejona86** [comment]
Flaked at https://travis-ci.org/grpc/grpc-java/jobs/140953978 (since rebuilt):
```
 io.grpc.netty.NettyTransportTest > ping_afterTermination FAILED
     java.lang.AssertionError: expected:<Status{code=UNAVAILABLE, description=null, cause=null}> but was:<Status{code=UNKNOWN, description=null, cause=java.nio.channels.ClosedChannelException}>
```

**Comment by buchgr** [comment]
Looking.... Also, Netty 4.1.2.Final comes with some additional hints of where the ClosedChannelException was triggered https://github.com/netty/netty/commit/e84567004324c0bfd04b0ca9e1e182ed89394b16#diff-8c9a7d1d9b0fdb6c5aef2794d05a9f88

**Comment by ejona86** [comment]
Seen again at https://travis-ci.org/grpc/grpc-java/jobs/143414728 (before the additional logging info) (since rebuilt)
```
io.grpc.netty.NettyTransportTest > ping_afterTermination FAILED
    java.lang.AssertionError: expected:<Status{code=UNAVAILABLE, description=null, cause=null}> but was:<Status{code=UNKNOWN, description=null, cause=java.nio.channels.ClosedChannelException}>
```

**Comment by buchgr** [comment]
Got one with logging enabled

```java
io.grpc.netty.NettyTransportTest > ping_afterTermination FAILED
    java.lang.AssertionError: java.nio.channels.ClosedChannelException
    	at io.netty.channel.AbstractChannel$AbstractUnsafe.write(...)(Unknown Source)
     expected:<Status{code=UNAVAILABLE, description=null, cause=null}> but was:<Status{code=UNKNOWN, description=null, cause=java.nio.channels.ClosedChannelException}>
```

Will look next week.

**Comment by buchgr** [comment]
ERIC ERIC !!1 I understand the race ... oh boy ... it's one hell of a race. 😆 .

**TL;DR;**
There are at _least_ two possible global execution orders, depending on whether `GracefulCloseCommand` and `SendPingCommand` are both flushed separately or in one go. We execute different error handling code in those two cases. In the first case, where there are two separate flushes, by the time `SendPingCommand` reaches the pipeline, the `NettyClientHandler` is no longer in the pipeline and [this code](https://github.com/grpc/grpc-java/blob/master/netty/src/main/java/io/grpc/netty/NettyClientTransport.java#L106) runs and the test succeeds. In the second case, the `NettyClientHandler` is still in the pipeline, and so `NettyClientHandler.sendPingFrame` is executed, which also fails due to a `ClosedChannelException`, but executes [different error handling code](https://github.com/grpc/grpc-java/blob/master/netty/src/main/java/io/grpc/netty/NettyClientHandler.java#L477).

**Execution Order 1 (Test succeeds)**
1. `WriteQueue.enqueue(new GracefulCloseCommand(...), true)`
2. **`WriteQueue.flush()`**
5. `ChannelPipeline.fireHandlerRemoved()`
3. `WriteQueue.enqueue(new SendPingCommand(...), true)`
4. **`WriteQueue.flush()`**

This first executes `NettyClientHandler.gracefulClose()`, which then calls `lifeCycleManager.notifyShutdown(Status)` and finally calls `ctx.close()`, which closes the channel AND enqueues the [inactive/unregister/**removeHandler**](https://github.com/netty/netty/blob/4.1/transport/src/main/java/io/netty/channel/AbstractChannel.java#L735)  logic as a task on the eventloop. The task gets executed _before_ the write queue's flush task, so by the time the ping command is written on the pipeline, the channel is closed, `NettyClientHandler` is no longer in the pipeline, and so  [this code](https://github.com/netty/netty/blob/4.1/transport/src/main/java/io/netty/channel/AbstractChannel.java#L791) fails the promise, which then executes [that code](https://github.com/grpc/grpc-java/blob/master/netty/src/main/java/io/grpc/netty/NettyClientTransport.java#L106). All good.

**Execution Order 2 (Test fails)**
1. `WriteQueue.enqueue(new GracefulCloseCommand(...), true)`
2. `WriteQueue.enqueue(new SendPingCommand(...), true)`
3. **`WriteQueue.flush()`**
5. `ChannelPipeline.fireHandlerRemoved()`

_Both_ commands are enqueued before flush. First, `NettyClientHandler.gracefulClose()` is run, which sets the status on the lifecycle manager and calls close, which closes the channel, but enqueues the [inactive/unregister/removeHandler](https://github.com/netty/netty/blob/4.1/transport/src/main/java/io/netty/channel/AbstractChannel.java#L735) logic as a task on the eventloop. So when we subsequently write the `SendPingCommand`, the `NettyClientHandler` is still there, calls `sendPingFrame`, which attempts to write a PING frame, but fails cause the channel is closed. However, the original promise from the write queue with the listener attached is not notified, cause it has [succeeded](https://github.com/grpc/grpc-java/blob/master/netty/src/main/java/io/grpc/netty/NettyClientHandler.java#L460) already. Instead, [this code](https://github.com/grpc/grpc-java/blob/master/netty/src/main/java/io/grpc/netty/NettyClientHandler.java#L477) runs and fails the ping callback with the `ClosedChannelException` as a cause, as it does not call `lifecycleManager.getShutdownStatus()` like the error handling code of the first promise...


Please don't tell me that this is flawed in some way and I am wrong 😅 

@ejona86 @nmittler

**Comment by nmittler** [comment]
ah, so we need to use the original promise all the way down to the wire, sounds like?

**Comment by ejona86** [comment]
So this is simply a case of needing to fix the [new ping's error handling logic](https://github.com/grpc/grpc-java/blob/master/netty/src/main/java/io/grpc/netty/NettyClientHandler.java#L477) to use [`statusFromFailedFuture()`](https://github.com/grpc/grpc-java/blob/v1.0.0/netty/src/main/java/io/grpc/netty/NettyClientTransport.java#L235).

We really need every future failure to go through that code path.

@nmittler, no I don't think the solution is to use the original promise. That still wouldn't communicate the correct status to the `Http2Ping`.

**Comment by buchgr** [comment]
@ejona86 yes this is also what I would suggest. I ll do a PR.

**Comment by ejona86** [comment]
It looks like `statusFromFailedFuture()` only needs the lifecycle manager. The synchronized block looks like it is [left-over](https://github.com/grpc/grpc-java/commit/1170afd16844bfe8654dd8634ff026d02e77baa0#diff-0121ed05a1dbf51add0d8961599a28c4) from a previous implementation.

It is very important to only call lifecycle manager from the netty event loop, because it is not thread-safe.

**Comment by ejona86** [comment]
@buchgr, a PR sounds great.
